src/neoxp/Node/NodeUtility.cs

```diff
@@ -208,6 +211,131 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+            string contractHash,
+            string rpcUri,
+            uint stateHeight)
+        {
+            if (!UInt160.TryParse(contractHash, out var _contractHash))
+            {
+                throw new ArgumentException($"Invalid contract hash: \"{contractHash}\"");    
+            }
+
+            if (!TransactionExecutor.TryParseRpcUri(rpcUri, out var uri))
+            {
+                throw new ArgumentException($"Invalid RpcUri value \"{rpcUri}\"");
+            }
+
+            using var rpcClient = new RpcClient(uri);
+            var stateAPI = new StateAPI(rpcClient);
+                
+            uint height = stateHeight;
+            if (height == 0)
+            {
+                var stateHeight_ = await stateAPI.GetStateHeightAsync();
+                if (stateHeight_.localRootIndex is null)
+                {
+                    throw new Exception("Null \"localRootIndex\" in state height response");
+                }
+                height = stateHeight_.localRootIndex.Value;
+            }
+
+            var stateRoot = await stateAPI.GetStateRootAsync(height);
+            var states = await rpcClient.ExpressFindStatesAsync(stateRoot.RootHash, _contractHash, new byte[0]);
+            var contractState = await rpcClient.GetContractStateAsync(contractHash).ConfigureAwait(false);
+
+            return (contractState, states);
+        }
+
+        public static int PersistContract(SnapshotCache snapshot, ContractState state,
+            IReadOnlyList<(string key, string value)> storagePairs, ContractCommand.OverwriteForce force)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_Contract).Add(state.Hash);
+            var localContract = snapshot.GetAndChange(key)?.GetInteroperable<ContractState>();
+
+            if (localContract is null)
+            {
+                // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+                // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+                // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+                // over keys that shouldn't exist for one of the contracts.
+                if (state.Id <= LastUsedContractId(snapshot))
+                {
+                    state.Id = GetNextAvailableId(snapshot);
+                }
+                else
+                {
+                    // Update available id such that a regular contract deploy will use the right next id;
+                    SetNextAvailableContractId(snapshot, state.Id + 1);
+                }
+
+                snapshot.Add(key, new StorageItem(state));
+            }
+            else
+            {
+                if (force == ContractCommand.OverwriteForce.None)
```

Comment:
We should check to see if the contract value matches the downloaded value, and only throw if there would be a need to overwrite

Reply:
We can validate `local contract state == downloaded contract state` and the same for all storage pairs, then give more specific errors ("Contract already exists - state and storage differs", ".. - storage differs", ".. - state differs" or "Contract already exists - aborting"). That would be friendlier to the user.